applications/FluidDynamicsApplication/custom_elements/d_vms.h

```diff
@@ -0,0 +1,435 @@
+
+    void GetValueOnIntegrationPoints(Variable<double> const& rVariable,
+                                     std::vector<double>& rValues,
+                                     ProcessInfo const& rCurrentProcessInfo) override;
+
+    void GetValueOnIntegrationPoints(Variable<array_1d<double, 6>> const& rVariable,
+                                     std::vector<array_1d<double, 6>>& rValues,
+                                     ProcessInfo const& rCurrentProcessInfo) override;
+
+    void GetValueOnIntegrationPoints(Variable<Vector> const& rVariable,
+                                     std::vector<Vector>& rValues,
+                                     ProcessInfo const& rCurrentProcessInfo) override;
+
+    void GetValueOnIntegrationPoints(Variable<Matrix> const& rVariable,
+                                     std::vector<Matrix>& rValues,
+                                     ProcessInfo const& rCurrentProcessInfo) override;
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    int Check(const ProcessInfo &rCurrentProcessInfo) override;
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    std::string Info() const override;
+
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override;
+
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+
+    ///@}
+
+protected:
+
+    ///@name Protected static Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    // Protected interface of FluidElement ////////////////////////////////////
+
+    void AddVelocitySystem(
+        TElementData& rData,
+        MatrixType& rLocalLHS,
+        VectorType& rLocalRHS) override;
+
+    void AddMassLHS(
+        TElementData& rData,
+        MatrixType& rMassMatrix) override;
+
+    // Implementation details of DVMS /////////////////////////////////////////
+
+    void AddMassStabilization(
+        TElementData& rData,
+        MatrixType& rMassMatrix);
+
+    void CalculateProjections(const ProcessInfo &rCurrentProcessInfo) override;
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
+private:
+
+    ///@name Static Member Variables
+    ///@{
+
+    constexpr static double mTauC1 = 8.0;
+    constexpr static double mTauC2 = 2.0;
+    constexpr static double mSubscalePredictionVelocityTolerance = 1e-14;
+    constexpr static double mSubscalePredictionResidualTolerance = 1e-14;
+    constexpr static unsigned int mSubscalePredictionMaxIterations = 10;
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    // Velocity subscale history, stored at integration points
+    std::vector< array_1d<double,Dim> > mPredictedSubscaleVelocity;
```

Comment:
is it deliberate to employ std::vector here?

Reply:
yes, it is variable size (num gauss points) and it is not expected to use it in combination with matrix operations. Which container would you suggest?

Reply:
i would have suggested the ublas::vector. You would need to do a resize instead of push back, but i think it is slightly faster. also it occupies less memory since it does not allocate "capacity" items